File: 0079-word-search/0079-word-search.cpp
int dirX[4] = {0, 0, 1, -1};
int dirY[4] = {1, -1, 0, 0};

class Solution {
public:
    int m, n;
    
    bool isValid(vector<vector<char>>& board, int i, int j, vector<vector<bool>>& visited) {
        return (i >= 0 && j >= 0 && i < m && j < n && !visited[i][j]);
    }
    
    bool dfs(vector<vector<char>>& board, int x, int y, vector<vector<bool>>& visited, string word, int index) {
        visited[x][y] = true;
        if (index == word.size()) {
            return true;
        }
        for (int i = 0; i < 4; i++) {
            int nx = x + dirX[i];
            int ny = y + dirY[i];
            if (isValid(board, nx, ny, visited) && board[nx][ny] == word[index]) {
                if (dfs(board, nx, ny, visited, word, index + 1)) {
                    return true;
                }
            }
        }
        visited[x][y] = false;
        return false;
    }
    
    bool exist(vector<vector<char>>& board, string word) {
        m = board.size(), n = board[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j] && board[i][j] == word[0]) {
                    if (dfs(board, i, j, visited, word, 1))
                        return true;
                }
            }
        }
        return false;
    }
};

File: 0079-word-search/0079-word-search.java
class Solution {
    
    public int[] dirs = new int[]{ 0, 1, 0, -1, 0 }; 
    public int m;
    public int n;
    
    public boolean exist(char[][] board, String word) {
        m = board.length;
        n = board[0].length;
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j] && board[i][j] == word.charAt(0)) {
                    if (solve(board, visited, i, j, 1, word)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }
    
    public boolean solve(char[][] board, boolean[][] visited, int x, int y, int it, String word) {
        visited[x][y] = true;
        if (it == word.length()) {
            return true;
        }
        for (int i = 0; i < 4; i++) {
            int nx = x + dirs[i], ny = y + dirs[i + 1];
            if (isValid(visited, nx, ny) && board[nx][ny] == word.charAt(it)) {
                if (solve(board, visited, nx, ny, it + 1, word)) {
                    return true;
                }
            }
        }
        visited[x][y] = false;
        return false;
    }
    
    public boolean isValid(boolean[][] visited, int x, int y) {
        return x >= 0 && y >= 0 && x < m && y < n && !visited[x][y];
    }
}